### PR TITLE
"pixelperfect" window aspect.

### LINKED ISSUE #1666: Add a project setting to force an integer scale for window stretch
**Describe the project you are working on:**
Low resolution 2D pixel art games.

**Describe the problem or limitation you are having in your project:**
Base engine stretch modes can preserve aspect ratio, but do not attempt to guarantee an even and square pixel grid.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
New project setting checkbox located at `display/window/stretch` labelled "Force Integer Scale" (or similar) which changes stretch behavior to attempt preservation of square pixels.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
I've already produced the feature for 3.2.X here: Yukitty/godot@42af4d50e696e790e33a0d9347fbe55e06da13e8
I am personally using this feature for all current and future 2D projects, as I do not produce or use high resolution art assets.

Ideal test project settings:
* `display/window/size/width` and `height` set to something low, like 256x224 or 320x240
(I am leaving `display/window/size/test_width` and `test_height` set to 1280x720 even in production)
* `display/window/size/resizable` On.
* `display/window/stretch/mode` 2d
* Use pixel art assets (Filter OFF) and pixel snap to produce a scene.

Test fiddling with these settings:
* `display/window/stretch/integer_scale` (the new feature)
* `display/window/stretch/aspect` (different behaviors per setting)

Demonstrated behavior while Integer Scale is active:
* If `display/window/stretch/aspect` is set to `Keep`, the game render will be letterboxed on ALL sides as the window expands, jumping up to the next available multiple of the base resolution _only_ when both the window's width and height are adequate to support it.
* If aspect is set to `Keep Width` or `Keep Height`, the game window will stretch vertically or horizontally, respectively, to almost fill the available space (discarding fractional pixels to preserve pixel accuracy). New letterbox bars will appear on the "kept" sides, which only jumps to resolution multiples.
* If aspect is set to `Expand`, the game's aspect ratio will expand in both directions simultaneously, which is another entirely new behavior.
* In all cases, if integer scale is checked, the pixels should be square, at all resolutions and window sizes, at all times. There may be minor rounding errors in my implementation, but the result is still notably better.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Almost. The primary blocker preventing a fully scripted solution from being viable is actually the font overdraw setting, which cannot be accessed outside of the SceneTree's root viewport stretch project settings, and is crucial for making dynamic fonts look good on a low resolution screen that's been scaled up.

Even avoiding that issue by using a pixel font, this behavior requires an obscure script and is not easily accessible to many users, as it's not a native setting of Viewports nor ViewportContainers that would display them.

There may be enough potential users to warrant a built-in simple setting, the same users of the Pixel Snap setting which currently work around this issue differently, by limiting the window size to specific game-enforced resolutions, or disabling the stretch feature entirely and writing it off as unusable.

**Is there a reason why this should be core and not an add-on in the asset library?:**
I can't currently fully replicate this behavior as an add-on due to the font overdraw, as described above.
Otherwise it may be viable, though it seems like an awfully minor feature to bury in the asset library. 🤔
It seems more suitably described as a base engine behavior rather than a key feature specific to any particular game or genre.

### PR DESCRIPTION
**Pull request by bfloch**
This mode makes sure that only integers are used as multiplier or divisor
for the viewport size, hence achieving "pixel perfect" scaling.
This works best with the `viewport` stretch mode.

As opposed to the other aspects this mode will allow for black borders
on all sides to satisfy the requirement.

It achieves a setup similar to https://alvarber.gitlab.io/pixel-perfection-in-godot.html
but just with a single setting as shown:
pixelperfect aspect

Since the godot build is broken for me in master/4.0 as reported by others in #43714 and #42348 I was not able to test this on master.

But I do have a sister patch for 3.2 in case you prefer this small, but useful feature for 3.x here:
https://github.com/StateOff/godot/commit/a63c016777f8b6aafcdc1a2ee7f2f08d59f4741c

Video showcasing the effect in 3.2:
https://i.imgur.com/kBb3Y49.mp4

https://user-images.githubusercontent.com/8281454/107992676-88b6ab00-701c-11eb-990d-467f8f346870.mp4

For convenience here is a stable 3.2.3 + patch Windows build for the curious without access to a compiler:
https://github.com/StateOff/godot/releases/tag/3.2.3-pixelperfect-0.1

Since I have limited access to platforms beyond Linux let me know if the patch needs more love.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/1666.*

### CODE REVIEW

**Comment by bfloch** [comment]
^ Force pushed std:: -> Math:: change.

**Comment by Calinou** [comment]
I don't think this should be a stretch aspect setting, as you may want to support multiple aspect ratios while having pixel-perfect scaling. Having a new project setting like https://github.com/godotengine/godot/pull/21446 did would be a better approach to this problem.

**Comment by bfloch** [comment]
Thanks for the feedback. It is a valid point - I did not anticipate the need for other aspects, frankly I have not seen this myself, but I understand that we should keep the engine as generic as possible and let the user define the use-cases.

Also I haven't visited the existing efforts and requests, so thanks for connecting the dots here.

I am ok with putting so more work into this as suggested, my only concern currently is that I can not run the master until some of the mentioned issues are resolved since I do not have a Mac or even Windows partition in my household at this point.

I am trying to suggest some patches which might ease the situation but there seems to be more work necessary to get 4.x somewhat stable on my platform at this point. So maybe it is best to postpone the work at bit.

**Comment by dalexeev** [comment]
This needs to be implemented either as a mode or as a separate property, but not as an aspect. See [Yukitty/godot-addon-integer_resolution_handler](https://github.com/Yukitty/godot-addon-integer_resolution_handler).

File: scene/main/window.cpp [hunk 1]
```diff
@@ -571,7 +571,20 @@ void Window::_update_viewport_size() {
...
+
+			if (video_mode >= desired_res) {
+				Size2 scale_size = video_mode / desired_res;
+				int scale_factor = Math::floor(std::min(scale_size.x, scale_size.y));
+				screen_size = desired_res * scale_factor;
+			} else {
+				Size2 scale_size = desired_res / video_mode;
+				int scale_factor = Math::ceil(std::max(scale_size.x, scale_size.y));
+				screen_size = desired_res / scale_factor;
+			}
```
Comment (akien-mga) [inline scene/main/window.cpp:586]: We have our own `MIN` and `MAX` macros which should be used instead of `std`.

**Comment by groud** [comment]
Aside from @akien-mga's comment, this looks good to me.

**Comment by Calinou** [comment]
See https://github.com/godotengine/godot/pull/46061#issuecomment-779402476. I think this is more of a stopgap than a definitive fix for https://github.com/godotengine/godot-proposals/issues/1666, so I'd prefer we spend more time working on a more durable solution.

**Comment by Calinou** [comment]
I have a PR that attempted to add integer scaling in a more flexible way, but didn't manage to finish it (and have no time to do so): https://github.com/godotengine/godot/pull/63206

**Comment by Calinou** [comment]
> If you don't have time might I take over that one if I find some time myself? 

Sure, go ahead :slightly_smiling_face:

**Comment by Calinou** [comment]
Superseded by https://github.com/godotengine/godot/pull/75784. Thanks for the contribution nonetheless!

**Pull request closed without merging**